phonon/update.py

```diff
@@ -234,5 +198,182 @@ def merge(self, update):
+        """
+        if block is None:
+            block = self.block
+
+        with self.ref.lock(block=block):
+            if self.ref.get_times_modified() > 0:
+                self._merge()
+            self.cache()
+            self.ref.increment_times_modified()
+
+        if self.init_cache:
+            self._clear()
+
+    def _execute(self, block=None):
+        """
+        Handles deciding whether or not to get the resource from redis. Also
+        implements merging cached records with this one (by implementing your
+        `merge` method). Increments the number of times this record was
+        modified if the cache method executes successfully (does not raise).
+        """
+        if block is None:
+            block = self.block
+
+        with self.ref.lock(block=block):
+            if self.ref.get_times_modified() > 0:
+                self._merge()
+            self.execute()
+
+
+class ConflictFreeUpdate(BaseUpdate):
+    """ 
+    Update class to be used when looking to minimize the locking of resources
+    during caching and execution of updates.  Accordingly, it should only be
+    used when you can ensure that you are only performing atomic operations on
+    both Redis and your database.
+
+    A default caching and merging method is provided which should be sufficient
+    when simply incrementing or decrementing fields. Otherwise, they must be
+    overwritten to suit your needs.
+    """
+
+    def __get_cached_state(self):
+        """
+        :rtype: dict
+        :returns: The update's state pulled from redis.
+        """
+        cached_state = self.process().client.hgetall(self.resource_id)
+        state = {}
+        for k, v in cached_state.items():
+            field, subfield = k.split('.') if '.' in k else (k, None)
+            if field == "doc":
+                continue
+            if subfield:
+                state.setdefault(field, {})[subfield] = v
+            else:
+                state[field] = v
+
+        empty_attributes = [k for k in self.state().keys() if k not in state]
+        for k in empty_attributes:
+            state[k] = self.clear().get(k) or {}
+
+        return state
+
+    def __get_cached_doc(self):
+        """
+        :rtype: dict
+        :returns: The update's doc pulled from redis.
+        """
+        return dict({(k.split(".")[1], v) for k, v in self.process().client.hgetall(self.resource_id).items() if "doc." in k})
+
+    def cache(self):
+        """
+        Caches the update's doc to redis within a hash named for the
+        resource_id and keyed by `doc.[key_name]`
+
+        This implementation should be used where data is only being
+        incremented or decremented.  Otherwise, overwrite to suite your
+        specific needs.
+        """
+        for k, v in self.doc.items():
+            self.process().client.hincrby(self.resource_id, "{}.{}".format("doc", k), int(v))
+
+    def cache_state(self):
```

Comment:
I think attempting to cache the state is causing more problems than it's worth. In the `BaseUpdate` class the intention of keeping state is to define a `__setstate__` method, making the instances pickle-able. This shouldn't be necessary in the conflict-free case since we'll never need to `pickle` in order to cache.

Let's just strip out the functionality that manages state for these updates and handle only the doc. I think this will simplify the code dramatically.